### PR TITLE
`!uqfinal` improvements

### PR DESCRIPTION
**Pull request by bradleysigma**
Now gives average scores needed when more than one assessment missing.
Accepts "%" in scores.  
Gives requirements for all passing grades.  
Reiterates inputs for both decimals and percentages.  
Names final assessment.  
Lists assessment if no marks given.  
image

### CODE REVIEW

File: uqcsbot/scripts/uqfinal.py [hunk 1]
```diff
@@ -48,24 +35,77 @@ def handle_uqfinal(command: Command):
...
+    total_deci = 0.0
+    results = []
+    for i, score_deci in enumerate(scores):
+        total_deci += score_deci * float(assessments[i]["weight"]) / 100
+        results.append(f"Inputted score of {round(score_deci * 100)}% for"
+                       f" {assessments[i]['taskName']} (weighted {assessments[i]['weight']}%)")
+    bot.post_message(command.channel_id, "\n".join(results))
+
+    # calculate remaining marks
+    remain_deci = 0.0
```
Comment (nicklambourne) [inline uqcsbot/scripts/uqfinal.py]: No part words, thanks

File: uqcsbot/scripts/uqfinal.py [hunk 2]
```diff
@@ -48,24 +35,77 @@ def handle_uqfinal(command: Command):
...
+    for i in range(len(scores), len(assessments)):
+        remain_deci += float(assessments[i]["weight"]) / 100
+
+    # calculate marks needed to achieve grades
+    message = []
+    for cutoff_deci, grade in [(0.5, 'four'), (0.65, 'five'), (0.75, 'six'), (0.85, 'seven')]:
```
Comment (nicklambourne) [inline uqcsbot/scripts/uqfinal.py]: Could this list become a constant?

File: uqcsbot/scripts/uqfinal.py [hunk 3]
```diff
@@ -48,24 +35,77 @@ def handle_uqfinal(command: Command):
...
 
-    needed = 50 - total
-    result = math.ceil(needed / float(course_info["assessment"][num_assessment - 1]["weight"]) * 100)
-    bot.post_message(command.channel_id, "You need to achieve at least " +
-                     str(result) +
-                     "% on the final exam.\n_Disclaimer: this does not take hurdles into account_\n_Powered by "
-                     "http://uqfinal.com_")
+    # if getting a four impossible
+    if not message:
+        message.append("I am a servant of the Secret Fire, wielder of the flame of Anor."
```
Comment (nicklambourne) [inline uqcsbot/scripts/uqfinal.py]: :oof:

**Comment by nicklambourne** [comment]
Also, the handle method desperately needs to be broken up into helper methods, if you're looking for things to do.

File: uqcsbot/scripts/uqfinal.py (hunk 1 above)
Comment (bradleysigma) [reply uqcsbot/scripts/uqfinal.py]: https://www.dictionary.com/browse/remain

File: uqcsbot/scripts/uqfinal.py (hunk 2 above)
Comment (bradleysigma) [reply uqcsbot/scripts/uqfinal.py]: Fixed.

**Pull request merged**